[GStreamer] Implement Speech Synthesis with Flite

shivamidow opened the pull request:
#### f99efce2948ad9f817f9e69e3272db98fafce4f0
<pre>
[GStreamer] Implement Speech Synthesis with Flite
<a href="https://bugs.webkit.org/show_bug.cgi?id=250628">https://bugs.webkit.org/show_bug.cgi?id=250628</a>

Reviewed by Philippe Normand.

This change implements the Speech Synthesis part of Web speech with Flite, an open-source
speech synthesis engine for ports using GStreamer. Once Flite converts texts to a wave,
GStreamer controls the playback of the wave sound in a GStreamer pipeline.
This patch only implements basic functionalities such as play, pause, resume, stop, and voice
change. Other fine controls, such as pitch, rate, etc., will be supported in the following changes.

This change passes 11 out of 13 tests in fast/speechsynthesis. We enable those tests in this patch,
skipping failed ones temporarily. We will fix the failures in a different bug.

Tests: LayoutTests/fast/speechsynthesis

Combined changes:
* LayoutTests/platform/glib/TestExpectations:
* Source/WebCore/PlatformGTK.cmake:
* Source/WebCore/PlatformWPE.cmake:
* Source/WebCore/platform/GStreamer.cmake:
* Source/WebCore/platform/PlatformSpeechSynthesizer.h:
* Source/WebCore/platform/graphics/gstreamer/GStreamerCommon.cpp: A new gst element added, &quot;webkitflitesrc&quot;.
(WebCore::registerWebKitGStreamerElements):
* Source/WebCore/platform/gstreamer/GUniquePtrFlite.h: Added.
* Source/WebCore/platform/gstreamer/PlatformSpeechSynthesizerGStreamer.cpp: Added. This file
  implements the playback of speech with given texts.
(WebCore::GstSpeechSynthesisWrapper::GstSpeechSynthesisWrapper):
(WebCore::GstSpeechSynthesisWrapper::~GstSpeechSynthesisWrapper):
(WebCore::GstSpeechSynthesisWrapper::pause):
(WebCore::GstSpeechSynthesisWrapper::resume):
(WebCore::GstSpeechSynthesisWrapper::speakUtterance):
(WebCore::GstSpeechSynthesisWrapper::cancel):
(WebCore::GstSpeechSynthesisWrapper::resetState):
(WebCore::GstSpeechSynthesisWrapper::handleMessage):
(WebCore::PlatformSpeechSynthesizer::create):
(WebCore::PlatformSpeechSynthesizer::PlatformSpeechSynthesizer):
(WebCore::PlatformSpeechSynthesizer::initializeVoiceList):
(WebCore::PlatformSpeechSynthesizer::pause):
(WebCore::PlatformSpeechSynthesizer::resume):
(WebCore::PlatformSpeechSynthesizer::speak):
(WebCore::PlatformSpeechSynthesizer::cancel):
(WebCore::PlatformSpeechSynthesizer::resetState):
* Source/WebCore/platform/gstreamer/WebKitFliteSourceGStreamer.cpp: Added. This file initializes
  a gstreamer source element with flite, registers voices.
(webkitFliteSrcConstructed):
(webkitFliteSrcStart):
(webkitFliteSrcStop):
(webkitFliteSrcCreate):
(webkitFliteSrcSetCaps):
(webkit_flite_src_class_init):
(fliteVoices):
(fliteVoice):
(ensureFliteVoicesInitialized):
(webKitFliteSrcSetUtterance):
* Source/WebCore/platform/gstreamer/WebKitFliteSourceGStreamer.h: Added.
* Source/WebKit/SourcesGTK.txt:
* Source/WebKit/SourcesWPE.txt:
* Source/WebKit/UIProcess/gstreamer/WebPageProxyGStreamer.cpp: Added.
(WebKit::WebPageProxy::didStartSpeaking):
(WebKit::WebPageProxy::didFinishSpeaking):
(WebKit::WebPageProxy::didPauseSpeaking):
(WebKit::WebPageProxy::didResumeSpeaking):
(WebKit::WebPageProxy::speakingErrorOccurred):
(WebKit::WebPageProxy::boundaryEventOccurred):
(WebKit::WebPageProxy::voicesDidChange):
* Source/WebKit/WebProcess/WebPage/WebPage.cpp: We don&apos;t install the speech synthesis client since
  we want to run gstreamer in Web Process.

(WebKit::m_appHighlightsVisible):
* Source/cmake/FindFlite.cmake: Added.
* Source/cmake/OptionsGTK.cmake:
* Source/cmake/OptionsWPE.cmake:
* Tools/Scripts/webkitpy/style/checker.py: Add WebKitFliteSourceGStreamer files for naming exceptions.

Canonical link: <a href="https://commits.webkit.org/259197@main">https://commits.webkit.org/259197@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/3643ffc5d7cacc4344018ae9a60958e9b4194958

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/104317 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/13402 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/37225 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/113530 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/173821 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/108246 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/14491 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/4317 "Built successfully") | [❌ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/96539 "Failed to checkout and rebase branch from PR 8677") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/112574 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/110085 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/11159 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/94223 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/36/builds/96539 "Failed to checkout and rebase branch from PR 8677") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/5/builds/107800 "Passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/93021 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/25834 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/36/builds/96539 "Failed to checkout and rebase branch from PR 8677") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/94320 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/6762 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/27191 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/92210 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/83/builds/4540 "Built successfully and passed tests") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/6895 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/3748 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/29977 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/12913 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/46749 "Passed tests") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/100891 "Built successfully") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/6337 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/8682 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/25037 "Passed tests") | 
| | | | | 
<!--EWS-Status-Bubble-End-->

philn commented on Source/WebCore/platform/gstreamer/PlatformSpeechSynthesizerGStreamer.cpp:
```diff
@@ -0,0 +1,231 @@
+    bool handleMessage(GstMessage*);
+
+    RefPtr<PlatformSpeechSynthesisUtterance> m_utterance;
+    PlatformSpeechSynthesizer* m_platformSynthesizer { nullptr };
```
Could this be some smart pointer?

shivamidow replied:
Yes. Done.

philn commented on Source/WebCore/platform/gstreamer/WebKitFliteSourceGStreamer.cpp:
```diff
@@ -0,0 +1,263 @@
+
+    GstAudioInfo info;
+    String text;
+    cst_voice* voice;
```
I suspect this should be stored in a GUniquePtr calling `delete_voice()` to release resources?

shivamidow replied:
Done. I applied  GUniquePtr to cst_voice*.

philn commented on Source/WebCore/platform/gstreamer/WebKitFliteSourceGStreamer.cpp:
```diff
@@ -0,0 +1,263 @@
+    ensureFliteVoicesInitialized();
+}
+
+static GstStateChangeReturn webkitFliteSrcChangeState(GstElement* element, GstStateChange transition)
```
Is this worth implementing only for logging purpose?

shivamidow replied:
Removed.

philn commented on Source/WebCore/platform/gstreamer/WebKitFliteSourceGStreamer.cpp:
```diff
@@ -0,0 +1,263 @@
+    if (!members->didLoadUtterance) {
+        members->didLoadUtterance = true;
+
+        cst_wave* wave = flite_text_to_wave(priv->text.utf8().data(), priv->voice);
```
When is `wave` freed?

shivamidow replied:
Deciding when we free cst_wave is tricky. Even before a WebKitFliteSrc instance is created, PlatformSpeechSynthesizer can request available voice lists. I made the cst_wave list static and applied a smart pointer to its elements.

philn commented on Source/WebCore/platform/gstreamer/WebKitFliteSourceGStreamer.cpp:
```diff
@@ -0,0 +1,263 @@
+        for (int i = 0; i < wave->num_samples; i++)
+            data[i * priv->info.channels] = wave->samples[i];
+
+        gst_adapter_push(members->adapter.get(), gst_buffer_ref(buf.get()));
```
use `buf.leakRef()` here, current code leaks I'm afraid.

shivamidow replied:
Fixed.

philn commented on Source/WebCore/platform/gstreamer/WebKitFliteSourceGStreamer.cpp:
```diff
@@ -0,0 +1,263 @@
+    baseSrcClass->create = GST_DEBUG_FUNCPTR(webkitFliteSrcCreate);
+    baseSrcClass->set_caps = GST_DEBUG_FUNCPTR(webkitFliteSrcSetCaps);
+
+    flite_init();
```
Maybe move this to a once init in the `constructed` func?

shivamidow replied:
O.K. Moved into ensureFliteVoicesInitialized()

philn commented on Source/WebCore/platform/gstreamer/WebKitFliteSourceGStreamer.cpp:
```diff
@@ -0,0 +1,263 @@
+        for (const cst_val* v = flite_voice_list; v; v = val_cdr(v)) {
+            cst_voice* voice = val_voice(val_car(v));
+            String voiceName = String::fromUTF8(voice->name);
+            voiceList.append(PlatformSpeechSynthesisVoice::create("http://festvox.org/flite/packed/flite-2.1/voices/cmu_us_"_s + voiceName + ".flitevox"_s, voiceName, "en-US"_s, false, false));
```
I wonder about other locales, can we load voices in other languages? is there something in the spec about that?

philn replied:
Also, how is flite loading its "voices"? Is it using the network? If so, that could be a WebProcess sandbox violation.

shivamidow replied:
Regarding other local languages, Maybe yes. Available voices provided by the project owner are English and Indian. I will be back to you with more info about that.

We can select a voice in two ways through flite_voice_select(). One is giving a voice name, and the other is giving a URL. I agree with your security concern, so the URI string is removed.

philn replied:
Ok, so there's no network usage at all, right? The voices are loaded from disk storage, correct?

mcatanzaro replied:
Also even if network usage was allowable, it would need to not use `http://`.

shivamidow replied:
@philn Correct. No network usage to get voices.
@mcatanzaro festvox.org does not support https, unfortunately. :(

philn commented on Source/cmake/OptionsGTK.cmake:
```diff
@@ -173,6 +173,7 @@ WEBKIT_OPTION_DEFAULT_PORT_VALUE(ENABLE_PERIODIC_MEMORY_MONITOR PRIVATE ON)
 WEBKIT_OPTION_DEFAULT_PORT_VALUE(ENABLE_POINTER_LOCK PRIVATE ON)
 WEBKIT_OPTION_DEFAULT_PORT_VALUE(ENABLE_SERVICE_WORKER PRIVATE ON)
 WEBKIT_OPTION_DEFAULT_PORT_VALUE(ENABLE_SHAREABLE_RESOURCE PRIVATE ON)
+WEBKIT_OPTION_DEFAULT_PORT_VALUE(ENABLE_SPEECH_SYNTHESIS PRIVATE ON)
```
Are we confident about enabling this by default?
What about WPE?

shivamidow replied:
Let's make it off until the feature is mature. I made it on due to layout tests enabled.

philn replied:
You can keep it as experimental feature, so it will be enabled for developer builds (bots, etc).

philn commented on Source/cmake/OptionsGTK.cmake:
```diff
@@ -173,6 +173,7 @@ WEBKIT_OPTION_DEFAULT_PORT_VALUE(ENABLE_PERIODIC_MEMORY_MONITOR PRIVATE ON)
 WEBKIT_OPTION_DEFAULT_PORT_VALUE(ENABLE_POINTER_LOCK PRIVATE ON)
 WEBKIT_OPTION_DEFAULT_PORT_VALUE(ENABLE_SERVICE_WORKER PRIVATE ON)
 WEBKIT_OPTION_DEFAULT_PORT_VALUE(ENABLE_SHAREABLE_RESOURCE PRIVATE ON)
+WEBKIT_OPTION_DEFAULT_PORT_VALUE(ENABLE_SPEECH_SYNTHESIS PRIVATE OFF)
```
`${ENABLE_EXPERIMENTAL_FEATURES}` and same for WPE please :)

shivamidow replied:
Done.

shivamidow committed:
[GStreamer] Implement Speech Synthesis with Flite
https://bugs.webkit.org/show_bug.cgi?id=250628

Reviewed by Philippe Normand.

This change implements the Speech Synthesis part of Web speech with Flite, an open-source
speech synthesis engine for ports using GStreamer. Once Flite converts texts to a wave,
GStreamer controls the playback of the wave sound in a GStreamer pipeline.
This patch only implements basic functionalities such as play, pause, resume, stop, and voice
change. Other fine controls, such as pitch, rate, etc., will be supported in the following changes.

This change passes 11 out of 13 tests in fast/speechsynthesis. We enable those tests in this patch,
skipping failed ones temporarily. We will fix the failures in a different bug.

Tests: LayoutTests/fast/speechsynthesis

Combined changes:
* LayoutTests/platform/glib/TestExpectations:
* Source/WebCore/PlatformGTK.cmake:
* Source/WebCore/PlatformWPE.cmake:
* Source/WebCore/platform/GStreamer.cmake:
* Source/WebCore/platform/PlatformSpeechSynthesizer.h:
* Source/WebCore/platform/graphics/gstreamer/GStreamerCommon.cpp: A new gst element added, "webkitflitesrc".
(WebCore::registerWebKitGStreamerElements):
* Source/WebCore/platform/gstreamer/GUniquePtrFlite.h: Added.
* Source/WebCore/platform/gstreamer/PlatformSpeechSynthesizerGStreamer.cpp: Added. This file
  implements the playback of speech with given texts.
(WebCore::GstSpeechSynthesisWrapper::GstSpeechSynthesisWrapper):
(WebCore::GstSpeechSynthesisWrapper::~GstSpeechSynthesisWrapper):
(WebCore::GstSpeechSynthesisWrapper::pause):
(WebCore::GstSpeechSynthesisWrapper::resume):
(WebCore::GstSpeechSynthesisWrapper::speakUtterance):
(WebCore::GstSpeechSynthesisWrapper::cancel):
(WebCore::GstSpeechSynthesisWrapper::resetState):
(WebCore::GstSpeechSynthesisWrapper::handleMessage):
(WebCore::PlatformSpeechSynthesizer::create):
(WebCore::PlatformSpeechSynthesizer::PlatformSpeechSynthesizer):
(WebCore::PlatformSpeechSynthesizer::initializeVoiceList):
(WebCore::PlatformSpeechSynthesizer::pause):
(WebCore::PlatformSpeechSynthesizer::resume):
(WebCore::PlatformSpeechSynthesizer::speak):
(WebCore::PlatformSpeechSynthesizer::cancel):
(WebCore::PlatformSpeechSynthesizer::resetState):
* Source/WebCore/platform/gstreamer/WebKitFliteSourceGStreamer.cpp: Added. This file initializes
  a gstreamer source element with flite, registers voices.
(webkitFliteSrcConstructed):
(webkitFliteSrcStart):
(webkitFliteSrcStop):
(webkitFliteSrcCreate):
(webkitFliteSrcSetCaps):
(webkit_flite_src_class_init):
(fliteVoices):
(fliteVoice):
(ensureFliteVoicesInitialized):
(webKitFliteSrcSetUtterance):
* Source/WebCore/platform/gstreamer/WebKitFliteSourceGStreamer.h: Added.
* Source/WebKit/SourcesGTK.txt:
* Source/WebKit/SourcesWPE.txt:
* Source/WebKit/UIProcess/gstreamer/WebPageProxyGStreamer.cpp: Added.
(WebKit::WebPageProxy::didStartSpeaking):
(WebKit::WebPageProxy::didFinishSpeaking):
(WebKit::WebPageProxy::didPauseSpeaking):
(WebKit::WebPageProxy::didResumeSpeaking):
(WebKit::WebPageProxy::speakingErrorOccurred):
(WebKit::WebPageProxy::boundaryEventOccurred):
(WebKit::WebPageProxy::voicesDidChange):
* Source/WebKit/WebProcess/WebPage/WebPage.cpp: We don't install the speech synthesis client since
  we want to run gstreamer in Web Process.

(WebKit::m_appHighlightsVisible):
* Source/cmake/FindFlite.cmake: Added.
* Source/cmake/OptionsGTK.cmake:
* Source/cmake/OptionsWPE.cmake:
* Tools/Scripts/webkitpy/style/checker.py: Add WebKitFliteSourceGStreamer files for naming exceptions.

Canonical link: https://commits.webkit.org/259197@main

mcatanzaro commented on Source/WebCore/platform/gstreamer/PlatformSpeechSynthesizerGStreamer.cpp:
```diff
@@ -0,0 +1,233 @@
+    GRefPtr<GstElement> m_src;
+};
+
+GstSpeechSynthesisWrapper::GstSpeechSynthesisWrapper(Ref<PlatformSpeechSynthesizer> synthesizer)
```
Should probably use `Ref<PlatformSpeechSynthesizer>&& synthesizer`?

philn replied:
Maybe! @shivamidow can you follow-up?

shivamidow replied:
Yes, I will.
